### DISCUSSION TITLE
开发模式下，monorepo项目，同时开启mfsu, extraBabelIncludes，修改公用子模块代码不会编译；

### DISCUSSION THREAD

**Discussion by codertony** [umijs/umi#11382] [Q&A]
## What happens?

monorepo项目，希望umi项目能够编译解析 包含@KM-R的依赖组件；

## Mini Showcase Repository(REQUIRED)
monorepo项目配置：
package.json
```
  "dependencies": {
    "umi": "^4.0.72",
    "@km-r/home-block": "workspace:*"
  },
```

umi配置：
```
{
  mfsu: {
    esbuild: true,
  },
  extraBabelIncludes: [/@km-r/],
}
```

## How To Reproduce

1.按照上面的配置同时开启mfsu和extraBabelIncludes，然后启动dev；
修改@km-r下的代码，编译不生效，刷新也没用；

2. 如果设置mfsu为false
修改@km-r下的代码，编译生效，且有热更新；

## Context

**Umi Version: 4.0.72
**Node Version: v16.16.0
**Platform: window10

**Comment by fz6m** [comment]
需要热更新的话，有两种解决办法。

一种是 mfsu 排除处理这个包（或者关闭 mfsu ），然后让 babel include 编译它，缺点是你要 watch 同步生成产物，这个时候读的是产物 js ：

```ts
  extraBabelIncludes: [
    'dep-name'
  ],
  mfsu: {
    exclude: ['dep-name']
  }
```

第二种是 [`monorepoRedirect`](https://umijs.org/docs/api/config#monoreporedirect) ，开启了这个选项之后就会自动把包的入口自动对准那个包的 src 文件夹里的 ts 源码，而不是产物，这样你就不需要 watch build 其他包了：

```ts
  monorepoRedirect: {}
```

注意这种方式你要加一个 tsconfig 的 paths 对准那个包路径才能获得 IDE 提示。

**Reply by stefango** [reply]
@fz6m hi，我试了下第二种方式 [monorepoRedirect](https://umijs.org/docs/api/config#monoreporedirect)，还是没支持热更新，期望效果是修改Copyright组件内容后，页面能自动更新。麻烦指导一下~
[demo](https://github.com/stefango/qiankun)
- 子应用组件 `examples/umi4/src/pages/index.tsx`
- 子应用配置 `examples/umi4/.umirc.ts`
- 其他子包组件 `examples/shared/libs/components/Copyright/Copyright.tsx`

```tsx
// Copyright.tsx
- return <div className={styles.copyright}>Copyright {new Date().getFullYear()}</div>;
+ return <div className={styles.copyright}>Updated Copyright {new Date().getFullYear()}</div>;
```

**Reply by fz6m** [reply]
在 `examples/shared` 这里面没有文件，需要给一个最小的复现。

**Reply by stefango** [reply]
@fz6m 抱歉，shared 文件被 git ignore 了，已更新（`examples/shared/lib` => `examples/shared/libs`）

**Reply by fz6m** [reply]
这个地方你的目录是 `libs` ，需要改一下，配错了：

```ts
  monorepoRedirect: {
    // 🟢
    srcDir: ['libs', 'src'],
  },
```

tsconfig 的类型提示：

```ts
  "shared/*": [
     "../shared/libs/*"
   ]
```

导入是直接从 `libs/*` 开始的，因为这个时候就默认认为 `libs` 是你的源码目录：

```ts
import { Copyright } from 'shared/components/Copyright';
```

另外：

1. `shared` 里的 `react` 依赖应该在 peerDeps 和 devDeps 里都放一份，而不是放到 deps 生产依赖里，虽然在 umi 4 里没有问题，但这里是不规范的，换别的框架或者工具可能就启动不起来了。

2. `shared` 里的 `react` 依赖是 17 ，而 umi 4 默认内置 react 是 18 ，这俩不一致，有风险，所以最好显示的在 umi 4 项目里安装 react 和 react-dom 统一大版本。

3. 不推荐在 `libs` 里写代码，一般我们都是写在 `src` 里，更推荐 `shared/src` 下写源码，这样就不需要配置 `srcDir` 了，默认就是指向 `src` 。

4. 不推荐使用深层导入 `shared/aaa/bbb` ，你可以在 `libs/index.ts` 创建一个出口汇集所有导出，这样就只需要写 `import { x } from 'shared'` （注意 tsconfig 需要去掉 `*` ）。

**Reply by stefango** [reply]
[demo](https://github.com/stefango/qiankun) 已更新，还有点问题
1. 更新 `examples/umi4/src/pages/index.tsx` 文件，主应用页面需要手动刷新页面才会更新（期望无需手动刷新，自动更新）
2. 更新 `examples/shared/libs/components/Copyright/Copyright.tsx`，主应用页面没更新（期望自动更新）

不知道是不是和这个 websocket 接口一直在 pending 有关
<img width="1559" alt="image" src="https://github.com/umijs/umi/assets/24286594/9598d1df-fe82-4cbd-b40c-236892f340b6">

**Reply by fz6m** [reply]
其他的问题自己排查下本地代理的问题，不要 proxy 127 和 localhost ，我本地是好的。

**Reply by stefango** [reply]
1. 感觉页面是有缓存，但是我看 `examples/umi4/node_modules/shared/src/components/Copyright/Copyright.tsx` 里是新的代码

<img width="904" alt="image" src="https://github.com/umijs/umi/assets/24286594/695bd9f4-732f-48b4-bfef-3b66a1289202">

<img width="1159" alt="image" src="https://github.com/umijs/umi/assets/24286594/5018b860-9106-4175-bc1c-da969a4816b4">

2. 接口 pending 应该不是代理问题，走代理的话 Remote Address 应该是 127.0.0.1:7890

![image](https://github.com/umijs/umi/assets/24286594/79bcc5d3-718c-4bbd-be5c-deb5bda46ef8)

<img width="675" alt="image" src="https://github.com/umijs/umi/assets/24286594/84466200-171c-4463-9514-8d7a22a02091">
